Gambot.Modules.Roll/DiceEvaluator.cs

```diff
@@ -0,0 +1,271 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Gambot.Modules.Roll
+{
+    public class DiceToken
```

Comment:
This class makes me uneasy. It's a roll-your-own-linked-list-of-unions-two-for-one-special. Is there a way this can be written in more idiomatic C#? Like, inheritance and generics built-in-collections and what have you